text/2022-11-08-font-compiler-ir.md

```diff
@@ -0,0 +1,190 @@
+
+# Basic structure
+
+```
+Source → IR → font binary
+            ^ one ir2font implementation
+       ^ many source2ir implementations
+```
+
+Source is one of
+
+- .glyphs
+- UFO (including many w/design space)
+- .ttx
+- nanoemoji config
+
+In future, sources could additionally include:
+
+- binary font
+  - Some contend if we support ttx we effectively support binary fonts
+- Unknown Future Design Format
+
+For every source we will have a dedicated conversion to IR, e.g. `glyphs2ir`, `ufo2ir`, etc in https://github.com/googlefonts/fontmake-rs.
+
+Let’s take `ufo2ir` as a concrete example, it would:
+
+- Depend on a UFO parser, likely https://github.com/linebender/norad
+  - Aside: norad uses quick-xml, we should try to stick to that elsewhere
+- Depend on the IR library
+
+# IR
+
+## Font sources
+
+Rough outline of common source structures
+
+UFO:
+
+- Designspace file maps UFO(s) into design space
+- UFO font has layers
+- Layers have glyphs
+
+Glyphs:
+
+- Font has glyphs
+- Glyphs have one layer per master
+- Plus user defined layers
+- Plus "special" layers that have attributes
+  - For example, "brace" layers to define intermediate masters for specific glyphs
+
+## Sketch
+
+Based on [ufo3](https://unifiedfontobject.org/versions/ufo3/), [babelfont-rs](https://github.com/simoncozens/babelfont-rs) and discussion in this repo. The
+structures sketched here in pseudo-Rust (based on multiple reports a similar format was helpful for colr). We deliberately have no concept of "master", we
+instead have a set of variable glyphs whose definition locations need not be identical. The examples are not exhaustive of all fields, they merely aspire to
+convey the general idea of a variable first, sparse first, structure broken apart in alignment with units of work for a parallel backend:
+
+```rust
+// data common to the entire designspace
+// written to disk in one file
+// some of https://unifiedfontobject.org/versions/ufo3/fontinfo.plist/ 
+struct Font
+  upem
+  axes: Vec<Axis>
+
+// TODO: do we really need "discrete" axes? - designspace 5.0
+enum AxisRange {
+  Continuous(min, max, default)
+  Discrete(Vec<f32>)  // e.g. Italic
+}
+struct Axis { tag, range: AxisRange, hidden }
+
+// Roughly
+type DesignSpaceLocation = HashMap<Tag, f32>
+
+// the complete variable definition of a single glyph
+// written to disk in one file per glyph
+// components point to other glyphs, in time expand to variable components
+// https://github.com/simoncozens/rust-font-tools/blob/main/babelfont-rs/src/glyph.rs 
+// an obvious unit of work for parallel execution
+// Note that this is sparse, not tied to masters, and the definition keys need not match glyph to glyph
+struct Glyph
+  name
+  definitions: Map<DesignSpaceLocation, GlyphInstance>
+
+// a specific instance of a glyph
+// https://unifiedfontobject.org/versions/ufo3/glyphs/glif/
+struct GlyphInstance
+  advance
+  outline: Vec<Shape>
+
+enum Shape
+  Contour(Contour)
+  Component(Component)
```

Comment:
In UFO you can indeed mix outlines with components. In retrospect I've found this needlessly complex. A model where you have a hard distinction between outlines and components as attributes of the glyph instance is much easier to work with. Something like:

```
struct GlyphInstance
  advance
  contours: Vec<Contour>
  components: Vec<Component>
```

Reply:
i agree it's better/easier to work with to have two separate lists of contours and components instead of a single vector of either contour or components.
The respective ordering of contours/components shouldn't matter in the end because in the output binary format mixed glyphs are going to be decomposed and the order of contours does not change the rendering (winding direction does).
Having said that, there's one thing that may care about the ordering of contours/components, and that's truetype instructions that refer to glyph point indices, see https://github.com/unified-font-object/ufo-spec/issues/169
Not sure if we care about representing those in the IR.

Reply:
TT can't mix outlines and components, so I think that's a non-issue.

Reply:
sure, in the binary TT file you can't mix, but in the source one may want to mix _and_ at the same time also have truetype instructions that refer to point indices. Even so, a compiler's composite glyph decomposition procedure could take care of remapping those indices accordingly. Anyway, this was all just hypothetical, am not suggesting we should use that as an argument for a mixed Vector[Union[Contour,Component]]

Reply:
Forgive my density but I'm not quite grasping why it is better to have separate lists, particularly in Rust where the enum for this is very natural?

Reply:
It makes it easy to determine if a glyph is a composite, and some other small things you do every now and then on glyphs.

Reply:
Does that mean if you wanted to roundtrip UFO:ir:UFO the combined list would be better (otherwise you can't restore the UFO properly), but if you mostly want to go source:ir:binary it's potentially slightly simpler to have two separate collections, and since we don't care about order they could be sets?

Reply:
Observation: RoboFont internally separates contours from components, and always writes the comonents last, regardless of the source UFO.

Sets are the wrong concept. While the order is irrelevant to what the final rendered shape looks like, it is still highly important when interpolating.

Reply:
Observation 2: both defcon and ufoLib2 separate contours from components. These are the two most use libraries in Python-land to write UFO, and both can't round-trip arbitrarily mixed contours/component sequences.

Reply:
> important when interpolating

Apparently I momentarily forgot variations. Some say they are important.

> both can't round-trip arbitrarily mixed contours/component sequences

This does feel like it's adding up to no significant benefit to mixing them together. I wonder why https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#outline is the way it is then.

Reply:
> I wonder why https://unifiedfontobject.org/versions/ufo3/glyphs/glif/#outline is the way it is then

Legacy. Insights progressed since. UFO is from 2001, 2? Fontographer did it that way. There's probably more to be found along those lines...

(The .glyphs format also seems to be able to arbitrarily mix contours and component references, FWIW.)

Reply:
Fixed in https://github.com/googlefonts/oxidize/pull/40.

Reply:
> .glyphs format also seems to be able to arbitrarily mix contours and component references

If Glyphs gains COLRv1 support and leverages the order of mixed contours and components we might find ourselves merging them back together. However, that's speculative so I think it's OK if we don't worry about it for now.